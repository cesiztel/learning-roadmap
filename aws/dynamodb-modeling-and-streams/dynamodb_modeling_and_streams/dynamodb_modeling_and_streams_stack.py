from aws_cdk import (
    Stack,
    aws_lambda_event_sources as eventsources,
    aws_dynamodb as dynamodb,
    aws_lambda as _lambda,
    Duration
)
from constructs import Construct


class DynamodbModelingAndStreamsStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Create the lambda trigger of the stream
        new_order_processor = _lambda.Function(
            scope=self,
            id="NewOrderProcessorStreamFunction",
            runtime=_lambda.Runtime.PYTHON_3_11,
            timeout=Duration.seconds(300),
            function_name="online-shop-new-order-processor-function",
            description="Function that process all the new orders of the shop",
            code=_lambda.Code.from_asset("assets/new_order_processor"),
            handler="index.lambda_handler",
            memory_size=128
        )


        # Explain the different types of StreamViewType
        data_store_table = dynamodb.Table(
            self, 
            "OnLineShopDataStoreTable",
            table_name="online-shop-data-store",
            partition_key=dynamodb.Attribute(
                name="PK",
                type=dynamodb.AttributeType.STRING
            ),
            sort_key=dynamodb.Attribute(
                name="SK",
                type=dynamodb.AttributeType.STRING
            ),
            stream=dynamodb.StreamViewType.NEW_IMAGE
        )

        # Set the lambda function as trigger of the function
        new_order_processor.add_event_source(
            eventsources.DynamoEventSource(
                table=data_store_table,
                starting_position=_lambda.StartingPosition.LATEST,
                metrics_config=_lambda.MetricsConfig(
                    metrics=[_lambda.MetricType.EVENT_COUNT]
                )
            )
        )
