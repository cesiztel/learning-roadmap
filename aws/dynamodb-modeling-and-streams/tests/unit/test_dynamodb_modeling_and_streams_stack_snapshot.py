import aws_cdk as cdk
from aws_cdk.assertions import Template

from dynamodb_modeling_and_streams.dynamodb_modeling_and_streams_stack import (
    DynamodbModelingAndStreamsStack,
)


# The snapshot parameter is injected by Pytest -- it's a fixture provided by
# syrupy, the snapshot testing library we're using:
# https://docs.pytest.org/en/stable/fixture.html
def test_matches_snapshot(snapshot):
    # Set up the app and resources in the other stack.
    app = cdk.App()

    # Create main DynamodbModelingAndStreamsStack 
    stack = DynamodbModelingAndStreamsStack(app, "dynamodb-modeling-and-streams")

    # Prepare the stack for assertions.
    template = Template.from_stack(stack)

    assert template.to_json() == snapshot
