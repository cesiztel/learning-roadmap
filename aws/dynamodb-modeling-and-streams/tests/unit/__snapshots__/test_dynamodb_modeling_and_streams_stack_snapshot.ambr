# serializer version: 1
# name: test_matches_snapshot
  dict({
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'NewOrderProcessorStreamFunction3E6FA205': dict({
        'DependsOn': list([
          'NewOrderProcessorStreamFunctionServiceRoleDefaultPolicy29216D80',
          'NewOrderProcessorStreamFunctionServiceRole20253325',
        ]),
        'Properties': dict({
          'Code': dict({
            'S3Bucket': dict({
              'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
            }),
            'S3Key': '25b58d503bfb33ff98254c84da4577337f983e5b1d5b3939ca6bf9b3fe6f2780.zip',
          }),
          'Description': 'Function that process all the new orders of the shop',
          'FunctionName': 'online-shop-new-order-processor-function',
          'Handler': 'index.lambda_handler',
          'MemorySize': 128,
          'Role': dict({
            'Fn::GetAtt': list([
              'NewOrderProcessorStreamFunctionServiceRole20253325',
              'Arn',
            ]),
          }),
          'Runtime': 'python3.11',
          'Timeout': 300,
        }),
        'Type': 'AWS::Lambda::Function',
      }),
      'NewOrderProcessorStreamFunctionDynamoDBEventSourcedynamodbmodelingandstreamsOnLineShopDataStoreTable0D14612CBE037DAB': dict({
        'Properties': dict({
          'BatchSize': 100,
          'EventSourceArn': dict({
            'Fn::GetAtt': list([
              'OnLineShopDataStoreTableF588F595',
              'StreamArn',
            ]),
          }),
          'FunctionName': dict({
            'Ref': 'NewOrderProcessorStreamFunction3E6FA205',
          }),
          'MetricsConfig': dict({
            'Metrics': list([
              'EventCount',
            ]),
          }),
          'StartingPosition': 'LATEST',
        }),
        'Type': 'AWS::Lambda::EventSourceMapping',
      }),
      'NewOrderProcessorStreamFunctionServiceRole20253325': dict({
        'Properties': dict({
          'AssumeRolePolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'sts:AssumeRole',
                'Effect': 'Allow',
                'Principal': dict({
                  'Service': 'lambda.amazonaws.com',
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'ManagedPolicyArns': list([
            dict({
              'Fn::Join': list([
                '',
                list([
                  'arn:',
                  dict({
                    'Ref': 'AWS::Partition',
                  }),
                  ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
                ]),
              ]),
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Role',
      }),
      'NewOrderProcessorStreamFunctionServiceRoleDefaultPolicy29216D80': dict({
        'Properties': dict({
          'PolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'dynamodb:ListStreams',
                'Effect': 'Allow',
                'Resource': '*',
              }),
              dict({
                'Action': list([
                  'dynamodb:DescribeStream',
                  'dynamodb:GetRecords',
                  'dynamodb:GetShardIterator',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::GetAtt': list([
                    'OnLineShopDataStoreTableF588F595',
                    'StreamArn',
                  ]),
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'PolicyName': 'NewOrderProcessorStreamFunctionServiceRoleDefaultPolicy29216D80',
          'Roles': list([
            dict({
              'Ref': 'NewOrderProcessorStreamFunctionServiceRole20253325',
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Policy',
      }),
      'OnLineShopDataStoreTableF588F595': dict({
        'DeletionPolicy': 'Retain',
        'Properties': dict({
          'AttributeDefinitions': list([
            dict({
              'AttributeName': 'PK',
              'AttributeType': 'S',
            }),
            dict({
              'AttributeName': 'SK',
              'AttributeType': 'S',
            }),
          ]),
          'KeySchema': list([
            dict({
              'AttributeName': 'PK',
              'KeyType': 'HASH',
            }),
            dict({
              'AttributeName': 'SK',
              'KeyType': 'RANGE',
            }),
          ]),
          'ProvisionedThroughput': dict({
            'ReadCapacityUnits': 5,
            'WriteCapacityUnits': 5,
          }),
          'StreamSpecification': dict({
            'StreamViewType': 'NEW_IMAGE',
          }),
          'TableName': 'online-shop-data-store',
        }),
        'Type': 'AWS::DynamoDB::Table',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
